File: packages/cli/tests/protocol-fixtures.test.ts
import { promises as fs } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";
import { getPaths } from "../src/protocol/paths.js";
import {
  assertWritableSchema,
  readArchiveManifest,
  readCache,
  readLock,
} from "../src/protocol/store.js";
import { makeTempHome } from "./helpers.js";

describe("desktop local protocol fixtures", () => {
  it("reads current lock and archive fixtures", async () => {
    const home = await makeTempHome();
    const paths = getPaths(home);
    await copyFixture("lock-v3-full.json", paths.agentLockFile);
    await copyFixture("archive-v1-full.json", paths.archiveManifestFile);

    const lock = await readLock(home);
    const manifest = await readArchiveManifest(home);

    expect(lock.version).toBe(3);
    expect(lock.skills.demo).toMatchObject({
      source: "owner/repo",
      sourceType: "github",
      sourceUrl: "https://github.com/owner/repo",
      ref: "main",
      skillPath: "skills/demo",
      commitSha: "abc123",
    });
    expect(manifest.version).toBe(1);
    expect(manifest.skills["demo-abc123"]).toMatchObject({
      archiveId: "demo-abc123",
      originalSkillId: "repo:owner/repo:demo",
      name: "demo",
      lockKey: "demo",
      archivedByVersion: "0.2.9",
    });
    expect(manifest.skills["demo-abc123"]?.apps.codex).toBe(true);
    expect(manifest.skills["demo-abc123"]?.lockEntry?.source).toBe("owner/repo");
  });

  it("applies defaults for minimal current fixtures", async () => {
    const home = await makeTempHome();
    const paths = getPaths(home);
    await copyFixture("lock-v3-minimal.json", paths.agentLockFile);
    await copyFixture("archive-v1-minimal.json", paths.archiveManifestFile);

    const lock = await readLock(home);
    const manifest = await readArchiveManifest(home);

    expect(lock.version).toBe(3);
    expect(lock.skills).toEqual({});
    expect(manifest.version).toBe(1);
    expect(manifest.skills).toEqual({});
  });

  it("tolerates legacy skills cache entries without apps", async () => {
    const home = await makeTempHome();
    const paths = getPaths(home);
    await fs.mkdir(path.dirname(paths.skillsCacheFile), { recursive: true });
    await fs.writeFile(
      paths.skillsCacheFile,
      JSON.stringify({
        skills: [
          {
            id: "ssot:legacy",
            name: "legacy",
            directory: "legacy",
            origin: "ssot",
            installedAt: 1000,
            updatedAt: 2000,
          },
        ],
      }),
    );

    const cache = await readCache(home);

    expect(cache.skills[0].apps).toBeUndefined();
  });

  it("refuses writes for future desktop protocol versions", async () => {
    const lockHome = await makeTempHome();
    await copyFixture("lock-v4-future.json", getPaths(lockHome).agentLockFile);
    await expect(async () => {
      assertWritableSchema(await readLock(lockHome), await readArchiveManifest(lockHome));
    }).rejects.toThrow("Lock file version 4 is newer than this CLI supports");

    const archiveHome = await makeTempHome();
    await copyFixture("archive-v2-future.json", getPaths(archiveHome).archiveManifestFile);
    await expect(async () => {
      assertWritableSchema(await readLock(archiveHome), await readArchiveManifest(archiveHome));
    }).rejects.toThrow("Archive manifest version 2 is newer than this CLI supports");
  });
});

function fixturePath(name: string): string {
  // import.meta.dirname (Node 21.2+) avoids Vite /@fs/ prefix on import.meta.url
  const dir = typeof import.meta.dirname === "string"
    ? import.meta.dirname
    : path.dirname(fileURLToPath(import.meta.url));
  return path.resolve(dir, "../../../fixtures/local-protocol", name);
}

async function copyFixture(name: string, destination: string): Promise<void> {
  await fs.mkdir(path.dirname(destination), { recursive: true });
  await fs.copyFile(fixturePath(name), destination);
}

File: src-tauri/tests/protocol_fixtures_test.rs
use skill_zoo_lib::persistence::{ArchiveManifest, ExternalImports};
use skill_zoo_lib::services::lock::SkillLock;
use std::fs;
use std::path::PathBuf;

#[test]
fn test_desktop_reads_current_protocol_fixtures() {
    let lock: SkillLock = read_fixture_json("lock-v3-full.json");
    let manifest: ArchiveManifest = read_fixture_json("archive-v1-full.json");
    let imports: ExternalImports = read_fixture_json("imports-v1-full.json");

    assert_eq!(lock.version, 3);
    let lock_entry = lock.skills.get("demo").expect("lock entry");
    assert_eq!(lock_entry.source.as_deref(), Some("owner/repo"));
    assert_eq!(lock_entry.source_type.as_deref(), Some("github"));
    assert_eq!(
        lock_entry.source_url.as_deref(),
        Some("https://github.com/owner/repo")
    );
    assert_eq!(lock_entry.branch.as_deref(), Some("main"));
    assert_eq!(lock_entry.skill_path.as_deref(), Some("skills/demo"));
    assert_eq!(lock_entry.commit_sha.as_deref(), Some("abc123"));

    assert_eq!(manifest.version, 1);
    let archived = manifest.skills.get("demo-abc123").expect("archive entry");
    assert_eq!(archived.archive_id, "demo-abc123");
    assert_eq!(archived.original_skill_id, "repo:owner/repo:demo");
    assert_eq!(archived.name, "demo");
    assert_eq!(archived.lock_key.as_deref(), Some("demo"));
    assert_eq!(archived.archived_by_version.as_deref(), Some("0.2.9"));
    assert_eq!(archived.apps.get("codex"), Some(&true));
    assert_eq!(
        archived
            .lock_entry
            .as_ref()
            .and_then(|entry| entry.source.as_deref()),
        Some("owner/repo")
    );

    assert_eq!(imports.version, 1);
    let import = imports
        .imports
        .get("external:demo-a1b2c3d4")
        .expect("external import");
    assert_eq!(import.directory, "skills/demo");
    assert_eq!(
        import.source_path,
        "/Users/example/private-skills/skills/demo"
    );
}

#[test]
fn test_desktop_applies_defaults_for_minimal_protocol_fixtures() {
    let lock: SkillLock = read_fixture_json("lock-v3-minimal.json");
    let manifest: ArchiveManifest = read_fixture_json("archive-v1-minimal.json");
    let imports: ExternalImports = read_fixture_json("imports-v1-minimal.json");

    assert_eq!(lock.version, 3);
    assert!(lock.skills.is_empty());
    assert_eq!(manifest.version, 1);
    assert!(manifest.skills.is_empty());
    assert_eq!(imports.version, 1);
    assert!(imports.imports.is_empty());
}

fn read_fixture_json<T: serde::de::DeserializeOwned>(name: &str) -> T {
    let content = fs::read_to_string(fixture_path(name)).expect("read fixture");
    serde_json::from_str(&content).expect("parse fixture")
}

fn fixture_path(name: &str) -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .expect("repo root")
        .join("fixtures")
        .join("local-protocol")
        .join(name)
}
